// fork from https://github.com/grpc/grpc/blob/master/examples/protos/helloworld.proto
// 
// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";
package protos;

service Ftp {
  // Hello to server
  rpc Hello (HelloRequest) returns (HelloResponse) {}
  
  // List remote directory
  rpc List (ListRequest) returns (stream ListResponse) {}
  
  // Pull a file from remote
  rpc Pull (PullRequest) returns (stream PullResponse) {}
  
  // Push a file to remote
  rpc Push (stream PushRequest) returns (PushResponse) {}
};

message HelloRequest {
  string message = 1;
};

message HelloResponse {
  string message = 1;
};

message ListRequest {
  File file = 1;
};

message ListResponse {
  File file = 1;
};

message PullRequest {
  File file = 1;
};

message PullResponse {
  FileSegment segment = 1;
};

message PushRequest {
  FileSegment segment = 1;
};

message PushResponse {
  int32 status = 1;
};

message FileSegment {
  string file_name = 1;
  bytes buf = 2;
  uint64 available_size = 3;
};

message File {
  enum Type {
    FILE = 0;
    DIRECTORY = 1;
  };
  string path = 1;
  Type type = 2;
  uint64 size = 3;
};

